pkg/runtime.v2/core/trainingruntime.go

```diff
@@ -0,0 +1,121 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	fwkcore "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/core"
+	fwkplugins "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/plugins"
+	idxer "github.com/kubeflow/training-operator/pkg/runtime.v2/indexer"
+)
+
+var (
+	errorNotFoundSpecifiedTrainingRuntime = errors.New("not found TrainingRuntime specified in TrainJob")
+)
+
+type TrainingRuntime struct {
+	framework *fwkcore.Framework
+	client    client.Client
+	scheme    *apiruntime.Scheme
+}
+
+var TrainingRuntimeGroupKind = schema.GroupKind{
+	Group: kubeflowv2.GroupVersion.Group,
+	Kind:  "TrainingRuntime",
+}.String()
+
+var _ runtime.Runtime = (*TrainingRuntime)(nil)
+
+var trainingRuntimeFactory *TrainingRuntime
+
+func NewTrainingRuntime(ctx context.Context, c client.Client, indexer client.FieldIndexer) (runtime.Runtime, error) {
+	if err := indexer.IndexField(ctx, &kubeflowv2.TrainJob{}, idxer.TrainJobTrainingRuntimeRefKey, idxer.IndexTrainJobTrainingRuntimes); err != nil {
+		return nil, fmt.Errorf("setting index on TrainingRuntime and ClusterTrainigRuntime for TrainJob: %w", err)
+	}
+	fwk, err := fwkcore.New(ctx, c, fwkplugins.NewRegistry(), indexer)
+	if err != nil {
+		return nil, err
+	}
+	trainingRuntimeFactory = &TrainingRuntime{
+		framework: fwk,
+		client:    c,
+		scheme:    c.Scheme(),
+	}
+	return trainingRuntimeFactory, nil
+}
+
+func (r *TrainingRuntime) NewObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob) ([]client.Object, error) {
+	var trainingRuntime kubeflowv2.TrainingRuntime
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: trainJob.Namespace, Name: trainJob.Spec.TrainingRuntimeRef.Name}, &trainingRuntime)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errorNotFoundSpecifiedTrainingRuntime, err)
+	}
+	return r.buildObjects(ctx, trainJob, trainingRuntime.Spec.Template, trainingRuntime.Spec.MLPolicy, trainingRuntime.Spec.PodGroupPolicy)
+}
+
+func (r *TrainingRuntime) buildObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob, jobSetTemplateSpec kubeflowv2.JobSetTemplateSpec,
+	mlPolicy *kubeflowv2.MLPolicy, podGroupPolicy *kubeflowv2.PodGroupPolicy) ([]client.Object, error) {
+	opts := []runtime.InfoOption{
+		runtime.WithLabels(jobSetTemplateSpec.Labels),
+		runtime.WithAnnotations(jobSetTemplateSpec.Annotations),
+		runtime.WithMLPolicy(mlPolicy),
+		runtime.WithPodGroupPolicy(podGroupPolicy),
+	}
+	for idx, rJob := range jobSetTemplateSpec.Spec.ReplicatedJobs {
+		if rJob.Replicas == 0 {
+			jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas = 1
+		}
+		replicas := jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas * ptr.Deref(rJob.Template.Spec.Completions, 1)
```

Comment:
Why are we doing this if Job replicas is always == 1 ?

Reply:
Does this mean that enforcing `.spec.replicatedJobs[*].replicas == 1` before we build the actual JobSet?

Reply:
Yes, I think, we should add this as part of validation to not allow users set values in `.spec.replicatedJobs[*].replicas == 1` in TrainingRuntime until we find a use-case when it is useful.

Reply:
That makes sense.
Let's validate this field in the CEL validation.
I will leave a comment on the CEL validation PR.

Reply:
As I mentioned in https://github.com/kubeflow/training-operator/pull/2248#discussion_r1799873876, I will move it to the CustomValidation and execute it during webhook.

Reply:
Done.